File: Python/61_rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return None
        last = head
        length = 1
        
        while (last.next):
            last = last.next
            length += 1
        
        k = k % length # e.g. if it rotate the same time as the length of list
        last.next = head # make it as circler list for now
        
        # last node is ( n - k -1)
        temp = head
        for _ in range(length - k -1):
            temp = temp.next
        
        answer = temp.next
        temp.next = None # cut off the last node
        
        
        return answer

File: Java/61_Rotate_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// Time O(N) Space O(1) 
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
     if (head==null||head.next==null) return head;
    ListNode dummy=new ListNode(0);
    dummy.next=head;
    ListNode fast=dummy,slow=dummy;

    int i;
    for (i=0;fast.next!=null;i++)//Get the total length 
    	fast=fast.next;
    
    for (int j=i-k%i;j>0;j--) //Get the i-n%i th node
    	slow=slow.next;
    
    fast.next=dummy.next; //Do the rotation
    dummy.next=slow.next;
    slow.next=null;
    
    return dummy.next;       
    }
}
